### PR TITLE
HPCC-22617 Fix SELF used in the rhs of an assignment inside a function definition

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
@shamser I have kept it as 3 commits to make it easier to review.
The first change moves existing code from HqlGram to the SelfReplacingTransformer class.  That is now created once per transform, rather than locally.  It also uses the lockTransformMutex() in the implicit in the base class of the QuickHqlTransformer, rather than explicitly calling it.  It encapsulates the code better, and is preparation for fixing the bug.
The second commit is small and simplifies the code.
The 3rd commit switches over to using the base functions of the QuickHqlTransformer - which removes some code, but more importantly means that function calls are now substituted correctly.

Examples edin18 and edin19 show the failure before, and the fix after.

**Comment by ghalliday** [comment]
This error seems to be unrelated to the PR:
```
Wutool error(s):
Cassandra:1) test: WuTool::testList (F) line: 1799 /mnt/disk1/home/vamosax/smoketest/PR-13185/HPCC-Platform/tools/wutool/wutool.cpp
Cassandra:assertion failed

```Repushed the branch to force the test to rerun.

File: ecl/hql/hqlgram2.cpp [hunk 1]
```diff
@@ -1710,14 +1653,138 @@ class SelfReferenceReplacer
...
-    OwnedHqlExpr ret = replacer.replaceExpression(rhs);
+bool HqlGram::checkAlreadyAssigned(const attribute & errpos, IHqlExpression * select)
+{
+    IHqlExpression * assigned = queryAlreadyAssigned(select);
+    if (!assigned)
+        return false;
+
+    StringBuffer s;
+    getFldName(assigned,s);
+    if (assigned == select)
```
Comment (shamser) [inline ecl/hql/hqlgram2.cpp]: Does this mean expression 'select' has been assigned to itself?

File: ecl/hql/hqlgram2.cpp [hunk 2]
```diff
@@ -1710,14 +1653,138 @@ class SelfReferenceReplacer
...
+    return curSelfReplacer ? curSelfReplacer->findAssignment(select) : nullptr;
+}
+
+/* All in parms will be consumed by this function */
+void HqlGram::addAssignment(attribute & target, attribute &source)
+{
+    OwnedHqlExpr targetExpr = target.getExpr();
+    OwnedHqlExpr srcExpr = source.getExpr();
+
+    if (!srcExpr) // something bad just happened.
```
Comment (shamser) [inline ecl/hql/hqlgram2.cpp]: Why isn't it reporting an error here?

File: ecl/hql/hqlgram2.cpp [hunk 3]
```diff
@@ -1710,14 +1653,138 @@ class SelfReferenceReplacer
...
+            addAssignall(targetExpr.getClear(), srcExpr.getClear(), target);
+        }
+        else
+            doAddAssignment(curTransform, targetExpr.getClear(), srcExpr.getClear(), target);
+    }
+    //else error occurred somewhere else
+}
+
+
+void HqlGram::addAssignment(const attribute & errpos, IHqlExpression * targetExpr, IHqlExpression * srcExpr)
```
Comment (shamser) [inline ecl/hql/hqlgram2.cpp]: Any reason not to common this up with the previous addAssigment implementation?

File: ecl/hql/hqlgram2.cpp [hunk 4]
```diff
@@ -12263,16 +12288,17 @@ IHqlExpression * HqlGram::endTransform(const attribute &errpos)
...
     curTransform = NULL;
     curTransformRecord.clear();
+    curSelfReplacer.clear();
     if (transformSaveStack.ordinality())
     {
         Owned<TransformSaveInfo> saved = &transformSaveStack.popGet();
         transformScope = saved->transformScope.getClear();
         curTransform = saved->curTransform.getClear();
         curTransformRecord.swap(saved->transformRecord);
+        curSelfReplacer.swap(saved->selfReplacer);
```
Comment (shamser) [inline ecl/hql/hqlgram2.cpp]: Why is swap used  here, rather than a simple set (like in lines 12298-12299)

**Review by shamser — COMMENTED** [review]
@ghalliday There's some comments requestion clarification.

**Review by shamser — APPROVED** [review]
@richardkchapman I've talked through the changes with Gavin.  Reviewed. Looks good.

**Comment by richardkchapman** [comment]
@ghalliday Did you want to squash before I merge?

**Comment by ghalliday** [comment]
@richardkchapman squashed

**Pull request merged**